internal/admin/handle_health_authority.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admin
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/google/exposure-notifications-server/internal/verification/database"
+	"github.com/google/exposure-notifications-server/internal/verification/model"
+)
+
+// HandleHealthAuthoritySave handles the create/update actions for health
+// authorities.
+func (s *Server) HandleHealthAuthoritySave() func(c *gin.Context) {
+	return func(c *gin.Context) {
+		var form healthAuthorityFormData
+		err := c.Bind(&form)
+		if err != nil {
+			ErrorPage(c, err.Error())
+			return
+		}
+
+		ctx := c.Request.Context()
+		m := TemplateMap{}
+
+		haDB := database.New(s.env.Database())
+		healthAuthority := &model.HealthAuthority{}
+		haID, err := strconv.ParseInt(c.Param("id"), 10, 64)
+		if err != nil {
+			ErrorPage(c, "Unable to parse `id` param")
+			return
+		}
+		if haID != 0 {
+			healthAuthority, err = haDB.GetHealthAuthorityByID(ctx, haID)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("error processing health authority: %v", err))
+				return
+			}
+		}
+		form.PopulateHealthAuthority(healthAuthority)
+
+		// Decide if update or insert.
+		updateFn := haDB.AddHealthAuthority
+		if haID != 0 {
+			updateFn = haDB.UpdateHealthAuthority
+		}
+		if err := updateFn(ctx, healthAuthority); err != nil {
+			ErrorPage(c, fmt.Sprintf("Error writing health authority: %v", err))
+			return
+		}
+
+		m.AddSuccess(fmt.Sprintf("Updated Health Authority '%v'", healthAuthority.Issuer))
+		m["ha"] = healthAuthority
+		m["hak"] = &model.HealthAuthorityKey{From: time.Now()} // For create form.
+		c.HTML(http.StatusOK, "healthauthority", m)
+	}
+}
+
+// HandleHealthAuthorityShow handles the show action for health authorities.
+func (s *Server) HandleHealthAuthorityShow() func(c *gin.Context) {
+	return func(c *gin.Context) {
+		ctx := c.Request.Context()
+		m := TemplateMap{}
+
+		healthAuthority := &model.HealthAuthority{
+			EnableStatsAPI: true, // default enabled.
+		}
+		if IDParam := c.Param("id"); IDParam == "0" {
```

Comment:
Does `idParam == ""` matter here?

Reply:
It's required to be a value by the router to reach this function. The router enforces a digit in the path.

This PR is just moving existing code around. To reduce the size of a diff and to reduce the chance of introducing a bug, I'm only moving code, no refactors or other changes in this PR.